### ISSUE TITLE
Saving Watch Later Playlist 

### ISSUE THREAD

**Issue by SheepDomination** [meeb/tubesync#439]
I want to add a _watch later_ playlist source. Although the directory which I want to save the media into is the same downloads folder which I setup TubeSync although I get this error; `You cannot specify a directory outside of the base directory (/downloads)` ?

And is there any means to which I can pick which videos from the watch later play list to download ?

**Comment by meeb** [comment]
The error is because you're trying to save downloads outside of the download root, check the source directory. Try setting a name like `watch-later`. If you've specified a full path like `/something/watch-later` this isn't inside the defined download root of `/download` and will throw the error you're seeing. All downloads must be within the `/downloads` root to make sense and be easily map-able via volumes from your container runtime.

**Comment by SheepDomination** [comment]
Although now I do not see any of the videos from the accounts watch later list ?

**Comment by meeb** [comment]
If you are trying to sync a private playlist (a playlist that requires you to be logged into a browser to view) you'll need to export your cookies and use those with tubesync - effectively you need to "log in" to YouTube within tubesync. See:

https://github.com/meeb/tubesync/blob/main/docs/using-cookies.md

**Comment by SheepDomination** [comment]
> If you are trying to sync a private playlist (a playlist that requires you to be logged into a browser to view) you'll need to export your cookies and use those with tubesync - effectively you need to "log in" to YouTube within tubesync. See:
> 
> https://github.com/meeb/tubesync/blob/main/docs/using-cookies.md

Do you know how to do this what is mentioned in the link; I'm using Edge and I've searched but can't find how ?

**Comment by meeb** [comment]
Click the "add playlist" button not "add channel".

**Comment by SheepDomination** [comment]
> Click the "add playlist" button not "add channel".

A source with this key already exists; that source is, YouTube playlist.

**Comment by meeb** [comment]
Then you've not deleted the old source yet.

**Comment by SheepDomination** [comment]
I'm getting 2 errors in the tasks.

**Comment by SheepDomination** [comment]
> That would imply that either your playlist URL doesn't work or your cookies weren't loaded properly. Check the logs for the word "cookies" and see if your `cookies.txt` file was properly loaded. If it was it would have been noted in the logs. This is in the container logs themselves, not in the web UI.

I don't see anything related to cookies.txt only, failed to load play list a few times among other things unrelated to cookies.txt ?

**Comment by meeb** [comment]
Then you may not have exported or saved your cookies.txt from youtube.com correctly. If working you'll get a note like

`[youtube-dl] using cookies.txt from: /some/path/to/cookies.txt`

in the container logs.

**Comment by SheepDomination** [comment]
`sync.youtube.YouTubeError: Failed to extract_info for "https://www.youtube.com/playlist?list=WL": No metadata was returned by youtube-dl, check for error messages in the logs above. This task will be retried later with an exponential backoff.`

This is the error to which the log is returning; same thing, its related to the cookies.txt file ?

**Comment by meeb** [comment]
Probably, yes. That error just means "yt-dlp tried to read the playlist and couldn't". Likely, the reason is because your cookies file isn't exported correctly or can't be read. If you don't see the `[youtube-dl] using cookies.txt from: ...` in your logs it won't work.

**Comment by SheepDomination** [comment]
Honestly, I don't know what is going wrong because I'm doing everything that I know correctly. I'm exporting out the cookies.txt file in Netscape format for only the URL specifically for accounts Watch Later list and all I get is the same error. And that error is, no metadata was returned by youtube-dl. 

I have an extension for my browser but the author of that extension just tells me, there are no problems; it's related to youtube-dl not understanding.  

I'm either going to never get this to work or someone hopefully can help me investigate what is causing the issue ?

**Comment by meeb** [comment]
Well, if you're confident that the cookies.txt file is correct then the only thing it can be is you've saved it in the wrong place or the file ownership or permissions are wrong. Paste the output of this command run on the host:

```bash
$ docker exec -ti tubesync ls -lat /config
```

**Comment by ShadowfootNZ** [comment]
This looks similar to the problem I have with the Watch Later playlist. #413 
I've just tried it again and it's still a problem. Looks like yt-dlp is still corrupting the cookies file so that it can't use it. There hasn't been an update in yt-dlp since then. I see someone has tested a patch to fix an issue with members-only cookies, so hopefully that will be the fix we need.

**Comment by meeb** [comment]
Thanks for looking into it. Hopefully if there is an issue here it'll get fixed with a new yt-dlp release. I don't have a Google account myself so I can't test things like watch later playlists so I'm reliant on accurate issues and logs to debug problems.

**Comment by SheepDomination** [comment]
> Well, if you're confident that the cookies.txt file is correct then the only thing it can be is you've saved it in the wrong place or the file ownership or permissions are wrong. Paste the output of this command run on the host:
> 
> ```shell
> $ docker exec -ti tubesync ls -lat /config
> ```

```
drwxr-xr-x 1 app  users     84 Dec  8 12:10 .
-rwxr-xr-x 1 app  users 380928 Dec  8 12:10 db.sqlite3
drwxr-xr-x 1 app  users     22 Dec  5 12:06 config
drwxr-xr-x 1 app  users     66 Dec  4 10:00 video
-rwxr-xr-x 1 app  users    825 Nov 27 14:18 compose.yaml
drwxr-xr-x 1 root root     254 Nov 26 09:28 ..
drwxr-xr-x 1 app  users      0 Nov 16 08:49 downloads
```
The output as you requested ?

**Comment by meeb** [comment]
That would imply your volumes defined in your `compose.yaml` are wrong and your `cookies.txt` file is in the wrong folder inside the container. What are your `volumes` in your `compose.yaml` for TubeSync?

**Comment by SheepDomination** [comment]
My volumes are: 
```
/volume1/docker/tubesync:/config
/volume1/docker/tubesync:/downloads
```

**Comment by meeb** [comment]
OK, so your volumes are very wrong. Try (after making sure the directories exist on the host):

```
/volume1/docker/tubesync/config:/config
/volume1/docker/tubesync/downloads:/downloads
```

And drop `cookies.txt` into your `config` directory. It's not being picked up in the container (and not working) because you have it at `/config/config/cookies.txt` and not `/config/cookies.txt`.

Re-using the same directory twice as a volume on the host usually makes little sense, certainly doesn't for TubeSync anyway, use two different directories. Also, don't put `config` on a network drive.

**Comment by SheepDomination** [comment]
> OK, so your volumes are very wrong. Try (after making sure the directories exist on the host):
> 
> ```
> /volume1/docker/tubesync/config:/config
> /volume1/docker/tubesync/downloads:/downloads
> ```
> 
> And drop `cookies.txt` into your `config` directory. It's not being picked up in the container (and not working) because you have it at `/config/config/cookies.txt` and not `/config/cookies.txt`.
> 
> Re-using the same directory twice as a volume on the host usually makes little sense, certainly doesn't for TubeSync anyway, use two different directories. Also, don't put `config` on a network drive.

The cookies.txt file is located in `/docker/tubesync/config/cookies.txt` 
Although 
![tubesyncerrro](https://github.com/meeb/tubesync/assets/6827459/fd104405-d753-4d69-ace5-71fc43dac5b2)

**Comment by meeb** [comment]
I meant inside the container the file needs to be at `/config/cookies.txt` which it currently is not. I don't wish to sound antagonistic here, but these are issues to do with the basics of running containers and not really anything to do with TubeSync. If you make the modifications I have suggested to your volumes it should work.

**Comment by SheepDomination** [comment]
> I meant inside the container the file needs to be at `/config/cookies.txt` which it currently is not. I don't wish to sound antagonistic here, but these are issues to do with the basics of running containers and not really anything to do with TubeSync. If you make the modifications I have suggested to your volumes it should work.

The cookies.txt file is placed in the directory path of /config/cookies.txt; although it still is not working ?

**Comment by SheepDomination** [comment]
```
drwxr-xr-x 1 app  users     42 Dec 12 01:53 .
-rwxr-xr-x 1 app  users 380928 Dec 12 01:53 db.sqlite3
-rwxr-xr-x 1 app  users   1344 Dec 12 01:53 cookies.txt
drwxr-xr-x 1 root root     254 Dec 11 14:44 ..
```
Result of the last command again. I'm thinking it may be the cookies file as it still doesn't work; but I don't know in what way to export the cookie file so that youtube-dl can read is correctly. As the `cookies.txt` file the format is Netscape and the domain is YouTube for the Watch Later list ?

**Comment by meeb** [comment]
... and the other question? Do you see a `[youtube-dl] using cookies.txt from: ...` in your logs? If you see that line your cookies are being found and loaded correctly. If that's still not working then it'll probably be your cookie file format (or URL?). There's loads of guides online about how to export cookies for `yt-dlp`.

You can also, if you like, directly test with `yt-dlp` on the command line to see if it works:

```bash
$ docker exec -ti tubesync yt-dlp --cookies /config/cookies.txt --simulate "https://www.youtube.com/playlist?list=WL"
```

The `--simulate` flag should test the request, but not download any videos.

Are you sure `https://www.youtube.com/playlist?list=WL` is the right URL? I don't have a YouTube account so I wouldn't know myself.

**Comment by ShadowfootNZ** [comment]
That is the correct url and it used to work.
How do we see what yt-dlp tries with cookies.txt? Is there a comprehensive logging we can do?

**Comment by meeb** [comment]
Sure, you can try:

```bash
$ docker exec -ti tubesync yt-dlp --verbose --cookies /config/cookies.txt --simulate "https://www.youtube.com/playlist?list=WL"
```

That'll produce a lot of logs. This is obviously just executing `yt-dlp` and nothing to do with TubeSync at this point. If you can get the above command to work with your `cookies.txt` (or find the new URL or whatever the issue is) then you can add it to TubeSync.

**Comment by SheepDomination** [comment]
> That is the correct url and it used to work. How do we see what yt-dlp tries with cookies.txt? Is there a comprehensive logging we can do?

Update on your results.

**Comment by meeb** [comment]
You can run that as well, along with anyone else trying to get `cookies.txt` files to work with `yt-dlp`.

**Comment by SheepDomination** [comment]
Same error as before;
```
[youtube:tab] Extracting URL: https://www.youtube.com/playlist?list=WL
[youtube:tab] WL: Downloading webpage
WARNING: [youtube:tab] YouTube said: The playlist does not exist.
ERROR: [youtube:tab] WL: YouTube said: The playlist does not exist.
```

**Comment by meeb** [comment]
If that URL works in your browser when you're logged in then I'd guess your `cookies.txt` file isn't valid. Given this is now a getting `yt-dlp` to work issue and nothing to do with TubeSync, nor do I use a `cookies.txt` file or log into YouTube I likely can't help much further other than point you in the direction of try updating your `cookies.txt` or try a different export format.

**Comment by SheepDomination** [comment]
> If that URL works in your browser when you're logged in then I'd guess your `cookies.txt` file isn't valid. Given this is now a getting `yt-dlp` to work issue and nothing to do with TubeSync, nor do I use a `cookies.txt` file or log into YouTube I likely can't help much further other than point you in the direction of try updating your `cookies.txt` or try a different export format.

What do you suggest ? 
TubeSync requires a Netscape format cookie which is what I'm exporting.

**Comment by meeb** [comment]
I guess try and get `yt-dlp` to work stand-alone with `cookies.txt` first? TubeSync literally just passes the path to the `cookies.txt` file to `yt-dlp` internally, TubeSync does no validation other than checking the file exists on disk and is readable. Look around for `yt-dlp` issues and see if anyone else has problems with the watch later playlist. This isn't anything TubeSync specifically is doing. But, logically, if that URL works when logged into a browser the only thing it can be is either an issue with `yt-dlp` itself (e.g. needs an update from upstream) or the `cookies.txt` permissions or format, there isn't anywhere else for there to be a problem. TubeSync does currently bundle the latest available `yt-dlp` release (2023.11.16 as of writing this comment).

**Comment by ShadowfootNZ** [comment]
This is with freshly extracted cookies, which I've seen working. 

`debug] Command-line config: ['--verbose', '--cookies', '/config/cookies.txt', '--simulate', 'https://www.youtube.com/playlist?list=WL']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2023.11.16 from yt-dlp/yt-dlp [24f827875] (pip)
[debug] Python 3.11.2 (CPython aarch64 64bit) - Linux-5.15.49-linuxkit-pr-aarch64-with-glibc2.36 (OpenSSL 3.0.11 19 Sep 2023, glibc 2.36)
[debug] exe versions: ffmpeg N-112750-g6d60cc7baf-20231114 (setts), ffprobe N-112750-g6d60cc7baf-20231114
[debug] Optional libraries: Cryptodome-3.19.0, brotli-1.1.0, certifi-2022.09.24, mutagen-1.47.0, requests-2.31.0, sqlite3-3.40.1, urllib3-2.1.0, websockets-12.0
[debug] Proxy map: {}
[debug] Request Handlers: urllib, requests
[debug] Loaded 1901 extractors
[youtube:tab] Extracting URL: https://www.youtube.com/playlist?list=WL
[youtube:tab] WL: Downloading webpage
[debug] [youtube:tab] Extracted SAPISID cookie
[youtube:tab] WL: Redownloading playlist API JSON with unavailable videos
[download] Downloading playlist: Watch later
[youtube:tab] Playlist Watch later: Downloading 19 items of 19
[download] Downloading item 1 of 19
`
After simulating the 19 downloads it finished with the following:
`[download] Finished downloading playlist: Watch later
`
I managed to repeat this several times. I then waited until Tubesync ran a task against an unlisted playlist. This was successful. However, repeating the --simulate I now get 
`[debug] Command-line config: ['--verbose', '--cookies', '/config/cookies.txt', '--simulate', 'https://www.youtube.com/playlist?list=WL']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2023.11.16 from yt-dlp/yt-dlp [24f827875] (pip)
[debug] Python 3.11.2 (CPython aarch64 64bit) - Linux-5.15.49-linuxkit-pr-aarch64-with-glibc2.36 (OpenSSL 3.0.11 19 Sep 2023, glibc 2.36)
[debug] exe versions: ffmpeg N-112750-g6d60cc7baf-20231114 (setts), ffprobe N-112750-g6d60cc7baf-20231114
[debug] Optional libraries: Cryptodome-3.19.0, brotli-1.1.0, certifi-2022.09.24, mutagen-1.47.0, requests-2.31.0, sqlite3-3.40.1, urllib3-2.1.0, websockets-12.0
[debug] Proxy map: {}
[debug] Request Handlers: urllib, requests
[debug] Loaded 1901 extractors
[youtube:tab] Extracting URL: https://www.youtube.com/playlist?list=WL
[youtube:tab] WL: Downloading webpage
WARNING: [youtube:tab] YouTube said: The playlist does not exist.
ERROR: [youtube:tab] WL: YouTube said: The playlist does not exist.
  File "/usr/local/lib/python3.11/dist-packages/yt_dlp/extractor/common.py", line 715, in extract
    ie_result = self._real_extract(url)
                ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/yt_dlp/extractor/youtube.py", line 4623, in wrapper
    info_dict = func(self, url, smuggled_data)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/yt_dlp/extractor/youtube.py", line 6583, in _real_extract
    self._extract_and_report_alerts(data, only_once=True)
  File "/usr/local/lib/python3.11/dist-packages/yt_dlp/extractor/youtube.py", line 786, in _extract_and_report_alerts
    return self._report_alerts(self._extract_alerts(data), *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/yt_dlp/extractor/youtube.py", line 783, in _report_alerts
    raise ExtractorError('YouTube said: %s' % errors[-1][1], expected=expected)
`

I see my cookies.txt has changed, and the change in behaviour differs depending on whether I'm using --simulate or not. 

1. I started with 167 lines in the exported cookies.txt file. 
2. When --simulate runs cookies.txt is updated to have 19 cookies (22 lines including one saying "# This file is generated by yt-dlp.  Do not edit.". 
3. When tubesync runs I have  I have 11 cookies. A key missing one is "LOGIN INFO". Also missing are: HSID, SSID, APISID, __Secure-1PAPISID, SID, __Secure-1PSID, SIDCC, __Secure-1PSIDCC.

Hopefully this is helpful to identify where cookies.text is being updated incorrectly.

**Comment by meeb** [comment]
Thanks for the details! Most helpful. Unfortunately though, TubeSync doesn't touch the `cookies.txt` file, at all. It literally just passes through the filename to `yt-dlp` over the API:

https://github.com/meeb/tubesync/blob/main/tubesync/sync/youtube.py#L35

So just this returns `{'cookiefile': '/path/to/cookies.txt'}` as arguments to `yt-dlp`. Any modifications to the file are going to be done internally, inside of `yt-dlp` and outside what I can look into or debug. It's possible that the handling of `--cookies` as a CLI argument is different to `cookiefile` as an API parameter, but I can't see where that might be. There aren't any other cookie related API parameters that I can see (other than `cookiesfrombrowser` which isn't useful here).

**Comment by SheepDomination** [comment]
> I guess try and get `yt-dlp` to work stand-alone with `cookies.txt` first? TubeSync literally just passes the path to the `cookies.txt` file to `yt-dlp` internally, TubeSync does no validation other than checking the file exists on disk and is readable. Look around for `yt-dlp` issues and see if anyone else has problems with the watch later playlist. This isn't anything TubeSync specifically is doing. But, logically, if that URL works when logged into a browser the only thing it can be is either an issue with `yt-dlp` itself (e.g. needs an update from upstream) or the `cookies.txt` permissions or format, there isn't anywhere else for there to be a problem. TubeSync does currently bundle the latest available `yt-dlp` release (2023.11.16 as of writing this comment).

When installing TubeSync by means of a container, how do I run programs such as `yt-dlp` and others that exist within the youtube-dl file system ?

**Comment by meeb** [comment]
> > I guess try and get `yt-dlp` to work stand-alone with `cookies.txt` first? TubeSync literally just passes the path to the `cookies.txt` file to `yt-dlp` internally, TubeSync does no validation other than checking the file exists on disk and is readable. Look around for `yt-dlp` issues and see if anyone else has problems with the watch later playlist. This isn't anything TubeSync specifically is doing. But, logically, if that URL works when logged into a browser the only thing it can be is either an issue with `yt-dlp` itself (e.g. needs an update from upstream) or the `cookies.txt` permissions or format, there isn't anywhere else for there to be a problem. TubeSync does currently bundle the latest available `yt-dlp` release (2023.11.16 as of writing this comment).
> 
> When installing TubeSync by means of a container, how do I run programs such as `yt-dlp` and others that exist within the youtube-dl file system ?

That's what this does

```bash
$ docker exec -ti tubesync yt-dlp ... args here ...
```

**Comment by meeb** [comment]
I'll close this for now. Feel free to comment if you are still experiencing issues.
